### PR TITLE
NGWMN-1717 SpringFox/Swagger (and prep for new APIs)

### PR DESCRIPTION
**Pull request by duselmann**
In order to cleanly swagger the API, I created a new JSON POJO and DSL builder following the lead of SpringFox DSL builder. Most of the code changes are to accommodate the new JsonData class structure.

SpringFox 2.7 did not have swagger examples implemented but it looks like 2.9 does.

### CODE REVIEW

File: src/main/java/gov/usgs/ngwmn/logic/WaterLevelStatistics.java [hunk 1]
```diff
@@ -78,14 +81,16 @@ public boolean doesThisMonthQualifyForStats(List<WLSample> monthSamples) {
 	private static final Logger LOGGER = LoggerFactory.getLogger(WaterLevelStatistics.class);
 	
 	public static enum MediationType {
-		BelowLand, 
-		AboveDatum;
+		NONE,
+		BelowLand,
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/ngwmn/logic/WaterLevelStatistics.java]: Is it worth refactoring the BelowLand and AboveDatum enum values to something like BELOW_LAND and ABOVE_DATUM?  It's a convention I've used/seen in other projects.

Comment (duselmann) [reply src/main/java/gov/usgs/ngwmn/logic/WaterLevelStatistics.java]: I did not know that. If that is true then I say, yes.

File: src/main/java/gov/usgs/wma/statistics/app/SwaggerConfig.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	@Autowired
+	private TypeResolver typeResolver;
+	
+	@Bean
+	public Docket qwPortalServicesApi() {
+		Docket docket = new Docket(DocumentationType.SWAGGER_2)
+//			.protocols(new HashSet<>(Arrays.asList("https")))
+//			.useDefaultResponseMessages(false)
+//			.host(configurationService.getSwaggerDisplayHost())
+//			.pathProvider(pathProvider())
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/statistics/app/SwaggerConfig.java]: Delete comments.

File: src/main/java/gov/usgs/wma/statistics/app/SwaggerParameterBuilder.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+				.required(true)
+				.build();
+	}
+	
+//	@ApiModel(description="CSV Descr",  value="JsonCSV")
+//	static class CSV {
+//		@ApiModelProperty(name="data1", notes="Data notes asdf", example = "CSV Data Example")
+//		@JsonProperty
+//		String data;
+//	}
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/statistics/app/SwaggerParameterBuilder.java]: Delete comments.

File: src/main/java/gov/usgs/wma/statistics/control/StatsService.java [hunk 4]
```diff
@@ -4,132 +4,139 @@
...
+import gov.usgs.wma.statistics.app.SwaggerConfig;
+import gov.usgs.wma.statistics.model.JsonData;
+import gov.usgs.wma.statistics.model.JsonDataBuilder;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
 
 @RestController
+@RequestMapping("/statistics")
 @CrossOrigin(origins = "*") // no credentials by default
+//@Api("StatisticsService")
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/statistics/control/StatsService.java]: Delete comments.

File: src/main/java/gov/usgs/wma/statistics/model/JsonOverall.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+/*
+"CALC_DATE": "2018-08-29",
+"LATEST_PCTILE": "0.31250"
+"LATEST_VALUE": "11.000",
+"MAX_VALUE": "1.000",
+"MEDIAN": "1.500",
+"MIN_VALUE": "43.000",
+"MAX_DATE": "2018-06-10T04:15:00-05:00",
+"MIN_DATE": "2005-06-10T04:15:00-05:00",
+*/
```
[the remark this answers is not in the document]
Comment (ssoper-usgs) [reply src/main/java/gov/usgs/wma/statistics/model/JsonOverall.java]: Or perhaps move this comment block into a javadoc comment over the constructor if the example values are helpful.

File: src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java [hunk 6]
```diff
@@ -157,8 +173,9 @@ public void test_monthlyStats_yearly_sameMonth() throws Exception {
 		List<Value> sorted = new LinkedList<>(samples);
 		StatisticsCalculator.sortByValueOrderAscending(sorted);
 
+		builder.mediation(MediationType.AboveDatum);
 		// we are not testing this method so mock it to return what we need
-		MonthlyStatistics<Value, MediationType> mockstats = new MonthlyStatistics<Value, MediationType>(MediationType.AboveDatum) {
+		MonthlyStatistics<Value> mockstats = new MonthlyStatistics<Value>(builder) {
 //			protected boolean doesThisMonthQualifyForStats(List<Value> monthSamples) {
 //				return monthSamples.size()>0;
 //			}
```
Comment (ssoper-usgs) [inline src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java]: Delete comments.

File: src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java [hunk 7]
```diff
@@ -337,9 +367,11 @@ public void testMostRecentProvistionalNONE_monathlyStats() {
...
 		StatisticsCalculator.sortByDateOrder(monthSamples);
-//		stats.overallStats(monthSamples, valueOrder, MediationType.AboveDatum);
 
-		Map<String, String> stat = stats.monthlyStats(monthSamples).get("3"); //MediationType.AboveDatum
+		boolean isCalc = stats.monthlyStats(monthSamples);
+		assertTrue(isCalc);
+		
+		Map<String, String> stat = builder.build().getMonthly().get("3").percentiles; // MediationType.AboveDatum
 
 //		assertNotNull(stat);
```
Comment (ssoper-usgs) [inline src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java]: Delete comments or un-comment if it's useful for the test

File: src/test/java/gov/usgs/wma/statistics/model/ValueTest.java [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+//		String bad = "2005-13-10T04:15:00-05:00";
+//		boolean actual = Value.checkBadTime(bad, 5, "test:N/A");
+//		assertTrue("No 13th month", actual);
+//	}
+//	@Test
+//	public void test_checkTime_day() {
+//		String bad = "2005-02-30T04:15:00-05:00";
+//		boolean actual = Value.checkBadTime(bad, 5, "test:N/A");
+//		assertTrue("Feb has < 30 days", actual);
+//	}
```
Comment (ssoper-usgs) [inline src/test/java/gov/usgs/wma/statistics/model/ValueTest.java]: Delete comments or un-comment.

File: src/main/java/gov/usgs/wma/statistics/app/SwaggerConfig.java (hunk 2 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/app/SwaggerConfig.java]: Yep, this was the first commit (ie not in the history) and I wanted to keep for a time the "documentation" of springfox for time. That said, it should go it not used.

File: src/main/java/gov/usgs/wma/statistics/app/SwaggerParameterBuilder.java (hunk 3 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/app/SwaggerParameterBuilder.java]: same as prior response

File: src/main/java/gov/usgs/wma/statistics/control/StatsService.java (hunk 4 above)
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/control/StatsService.java]: Same as prior response.

File: src/main/java/gov/usgs/wma/statistics/model/JsonOverall.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/java/gov/usgs/wma/statistics/model/JsonOverall.java]: It should be javadoc. This was a quick paste to ensure I could recall the current response to ensure the new class matched fields.

File: src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java (hunk 6 above)
Comment (duselmann) [reply src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java]: Oddly, I think the next PR will clear this.

File: src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java (hunk 7 above)
Comment (duselmann) [reply src/test/java/gov/usgs/wma/statistics/logic/MonthlyStatisticsTest.java]: agreed

File: src/test/java/gov/usgs/wma/statistics/model/ValueTest.java (hunk 8 above)
Comment (duselmann) [reply src/test/java/gov/usgs/wma/statistics/model/ValueTest.java]: agreed

**Pull request merged**